torch/csrc/autograd/custom_function.cpp

```diff
@@ -102,8 +102,14 @@ variable_list _wrap_outputs(const variable_list &input_vars,
   int num_diff_outputs = 0;
 
 
+  uint32_t num_tensors = 0;
   for (auto i = 0; i < num_outputs; ++i) {
     Variable var = raw_outputs[i];
+    if (!var.defined()) {
```

Comment:
What if the user returned an undefined Tensor on which he wants history to be tracked?

Reply:
The thing is that this code is in common for both python and c++
But I could see users trying to track gradients for their custom objects and using undefined Tensors as a "cheap" object to give to the autograd. Not that I would recommend doing that :p 

I think that if you only want to make the python version work for checkpoint, it will be simpler for sure to do a wrapper on checkpoint.

Otherwise, I think we will have to update this function to take list of optional Tensor to be able to properly mark the outputs that should not be considered and/or use mark_non_differentiable. But that might have more perf implications...

Reply:
> But I could see users trying to track gradients for their custom objects and using undefined Tensors as a "cheap" object to give to the autograd. Not that I would recommend doing that :p

Hmm I'm still not sure how an undefined Tensor can be used to track autograd history? If the output of a function is an undefined tensor, can't it be ignored for autograd purposes (as we do here)? 

> I think that if you only want to make the python version work for checkpoint, it will be simpler for sure to do a wrapper on checkpoint.

I guess the goal here is to support this for python custom functions in general and not just checkpoint. In particular classes like these: https://github.com/pytorch/pytorch/blob/master/torch/distributed/pipeline/sync/checkpoint.py#L249.

Reply:
My question is that if we only want to support that for python function, is there a way to make this change python only? That will make things easier to review and less likely to break the c++ Function in a subtle way.